lib/wasmex/component_instance.ex

```diff
@@ -0,0 +1,36 @@
+defmodule Wasmex.Component.Instance do
+  defstruct store_resource: nil,
+            instance_resource: nil,
+            # The actual NIF store resource.
+            # Normally the compiler will happily do stuff like inlining the
+            # resource in attributes. This will convert the resource into an
+            # empty binary with no warning. This will make that harder to
+            # accidentally do.
+            reference: nil
+
+  def __wrap_resource__(store_resource, instance_resource) do
+    %__MODULE__{
+      store_resource: store_resource,
+      instance_resource: instance_resource,
+      reference: make_ref()
+    }
+  end
+
+  def new(store_or_caller, component) do
+    %{resource: store_or_caller_resource} = store_or_caller
+    %{resource: component_resource} = component
+
+    case Wasmex.Native.component_instance_new(store_or_caller_resource, component_resource) do
+      {:error, err} -> {:error, err}
+      resource -> {:ok, __wrap_resource__(store_or_caller_resource, resource)}
+    end
+  end
+
+  def call_function(
+        %__MODULE__{store_resource: store_resource, instance_resource: instance_resource},
+        function,
+        args
+      ) do
+    {:ok, Wasmex.Native.exec_func(store_resource, instance_resource, function, args)}
```

Comment:
* I think it's good to have `exec_fun` separate from `instance_call_exported_function`, but both are very similar things. I would rename both to have the consistent naming (e.g. `instance_call_function` and `component_call_function`). What do you think? No need to rename `instance_call_exported_function` in this PR (can be a follow-up) but the new call-function name can starts in a good place when we just introduce it 🤔

Reply:
Agreed, renaming this was one of the next things on my list. I'd like to keep it consistent as well, at first I was just thinking I'd rename mine to `component_instance_call_exported_function` but that is a mouthful :) I'm good with `component_call_function`